p2p/transport.go

```diff
@@ -0,0 +1,579 @@
+}
+
+// transportLifecycle bundles the methods for callers to control start and stop
+// behaviour.
+type transportLifecycle interface {
+	Close() error
+	Listen(NetAddress) error
+}
+
+// ConnFilterFunc to be implemented by filter hooks after a new connection has
+// been established.
+type ConnFilterFunc func(map[string]net.Conn, net.Conn) error
+
+// PeerFilterFunc to be implemented by filter hooks after a new Peer has been
+// fully setup.
+type PeerFilterFunc func(map[ID]Peer, Peer) error
+
+// IPResolver is a behaviour subset of net.Resolver.
+type IPResolver interface {
+	LookupIPAddr(context.Context, string) ([]net.IPAddr, error)
+}
+
+// ConnDuplicateIPFilter resolves and keeps all ips for an incoming connection
+// and refuses new ones if they come from a known ip.
+func ConnDuplicateIPFilter(resolver IPResolver) ConnFilterFunc {
+	im := map[string][]net.IP{}
+
+	return func(cs map[string]net.Conn, c net.Conn) error {
+		// Remove cached ips for dropped conns.
+		for addr := range im {
+			if _, ok := cs[addr]; !ok {
+				delete(im, addr)
+			}
+		}
+
+		// Resolve ips for incoming conn.
+		host, _, err := net.SplitHostPort(c.RemoteAddr().String())
+		if err != nil {
+			return err
+		}
+
+		addrs, err := resolver.LookupIPAddr(context.Background(), host)
+		if err != nil {
+			return err
+		}
+
+		newIPs := []net.IP{}
+
+		for _, addr := range addrs {
+			newIPs = append(newIPs, addr.IP)
+		}
+
+		for _, knownIPs := range im {
+			for _, known := range knownIPs {
+				for _, new := range newIPs {
+					if new.Equal(known) {
+						return &ErrRejected{
+							conn:        c,
+							err:         fmt.Errorf("IP<%v> already connected", new),
+							isDuplicate: true,
+						}
+					}
+				}
+			}
+		}
+
+		im[c.RemoteAddr().String()] = newIPs
+
+		return nil
+	}
+}
+
+// MultiplexTransportOption sets an optional parameter on the
+// MultiplexTransport.
+type MultiplexTransportOption func(*MultiplexTransport)
+
+// MultiplexTransportConnFilters sets the filters for rejection new connections.
+func MultiplexTransportConnFilters(
+	filters ...ConnFilterFunc,
+) MultiplexTransportOption {
+	return func(mt *MultiplexTransport) { mt.connFilters = filters }
+}
+
+// MultiplexTransportFilterTimeout sets the timeout waited for filter calls to
+// return.
+func MultiplexTransportFilterTimeout(
+	timeout time.Duration,
+) MultiplexTransportOption {
+	return func(mt *MultiplexTransport) { mt.filterTimeout = timeout }
+}
+
+// MultiplexTransportPeerFilters sets the filters for rejection of new peers.
+func MultiplexTransportPeerFilters(
+	filters ...PeerFilterFunc,
+) MultiplexTransportOption {
+	return func(mt *MultiplexTransport) { mt.peerFilters = filters }
+}
+
+// MultiplexTransport accepts and dials tcp connections and upgrades them to
+// multiplexed peers.
+type MultiplexTransport struct {
+	listener net.Listener
+
+	acceptc chan accept
+	closec  chan struct{}
+
+	// Lookup table for duplicate ip and id checks.
```

Comment:
can't we just use the existing PeerSet for this purpose?

Reply:
Great suggestion, will investigate!